plugins/magic.py

```diff
@@ -1,39 +1,67 @@
-"""!magic <search term> return the first magic card with name containing <search term>"""
+"""!magic <search term> return magic card specified by the search term"""
 
+import math
 import re
 import random
 from urllib import quote
 
 import requests
 
-def magic(searchterm):
-    quotedsearchterm = quote(searchterm)
-    url = "https://api.deckbrew.com/mtg/cards?name={}"
-    url = url.format(quotedsearchterm)
-    results = requests.get(url).json()
+filterlist = ["type","supertype","subtype","name","oracle","set","rarity","set","color","multicolor","m","format","status"]
 
-    if not results:
+def search(searchterm):
+    filters = getfilters(searchterm)
+    if filters:
+        search = reduce(lambda string, term: string + ("" if not string else "&") + "{0}={1}".format(term[0].strip(),quote(term[1].strip())), filters, "")
+        url = "https://api.deckbrew.com/mtg/cards?{}".format(search)
+        result = getresult(url)
+    else:
+        #if no filters found we default to name search
+        url = "https://api.deckbrew.com/mtg/cards?name={}".format(quote(searchterm))
+        results = requests.get(url).json()
+        if not results:
+            return "no cards found"
+        #pick a random card from the matches in case we don't find a match
+        result = random.choice(results)
+        for card in results:
+            if searchterm.lower() == card["name"].lower():
+                #exact match found, pick a random edition with valid image
+                result = card
+                break
+    
+    if not result:
         return "no cards found"
+    return getimage(result)
 
-    for card in results:
-        if searchterm.lower() == card["name"].lower():
-            #exact match found, pick a random edition with valid image
-            return getimage(card)
-        
-    #no exact match found, pick a random card from the matches
-    card = random.choice(results)
-    return getimage(card)
-
-def cardname(searchterm):
-    quotedsearchterm = quote(searchterm)
-    url = "https://api.deckbrew.com/mtg/cards?name={}"
-    url = url.format(quotedsearchterm)
-    return magic(url)
+#find number of results and get a random result
+def getresult(url):
+    lo = 1
+    hi = 2
+    while requests.get(url + "&page={}".format(hi-1)).json():
+        lo = hi
+        hi *= 2
+    while lo < hi:
+        page = int(math.ceil(float(lo + hi)/float(2)))
+        if requests.get(url + "&page={}".format(page-1)).json():
+            lo = page
+        else:
+            hi = page - 1
+    results = 100*(lo-1) + len(requests.get(url + "&page={}".format(lo-1)).json())
```

Comment:
We can get this from the number of items on the first page, in case the number per page changes.